In [1]:
from sklearn.datasets import fetch_california_housing
X, y = fetch_california_housing(as_frame=True, return_X_y=True)

In [2]:
y

0        4.526
1        3.585
2        3.521
3        3.413
4        3.422
         ...  
20635    0.781
20636    0.771
20637    0.923
20638    0.847
20639    0.894
Name: MedHouseVal, Length: 20640, dtype: float64

# Practice Quetions

Total # of ques: 8

**Instruction:** Based on california housing dataset answer the Questions that follow.

## (1) What is the shape of feature matrix for the dataset?

a. (20640, 8)

b. (2064, 8)

c. (20610, 8)

d. (20600, 8)

**Correct Answer: a**

**Solution**

In [3]:
X.shape

(20640, 8)

## (2) How many missing values are present in the dataset?

a. 100

b. 157

c. 0

d. 87

**Correct answer: c**

**Solution**

In [4]:
X.isna().sum().sum()

0

**Instructions**: For Que 3 to 6, split the california housing dataset into train and test sets.

[Hint: us appropriate API with random_state=0, shuffle=False and test_size=0.2]

## Que 4

What is the value of Population in the first sample of training set?

a) 2401

b) 1700

c) 322

d) 157

**Correct Anwswer: c**



In [5]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test =train_test_split(X,y,random_state=0, shuffle=False,test_size=0.2)

## Que 5

What is the mean of output label in training set?

a) 1.98

b) 1.96

c) 2.04

d) 2.02

**Correct Answer:d**


In [8]:
y_train.mean()

2.020670313105598

## Que 6

What is the median of output label in test set?

a) 2.02

b) 1.96

c) 2.14

d) 2.34

**Correct Answer: c**

In [9]:
y_test.median()

2.1475

## **Que 7**

If we use `StandardScaler` for preprocessing and `LinearRegression` for fitting the model with training set obtained in previous questions, what are the coefficients obtained in the predicted model?

**Note:** Use methods and objects with default parameters.

a. [ 0.7296 ,  0.2187, -0.3655,  0.4056,   -0.0045  , -0.0493, -0.9998, -0.8705  ]

b. [ 0.9296 ,  0.2287, -0.3655,  0.4056,   -0.0045  , -0.0393, -0.8998, -0.8705  ]

c. [ 0.2961,  0.1875, -0.6552,  0.30569623,   -0.0045  , -0.0393, -0.8998, -0.8705  ]

d. [ 0.8381,  0.1203, -0.3213,  0.3645, -0.0017,-0.0452, -0.8799, -0.8263]

**Correct Answer : d**

**Solution**

In [10]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline

In [16]:
pipe=Pipeline(steps=[('scaler',StandardScaler()),("LR",LinearRegression())])

In [17]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('scaler', StandardScaler()), ('LR', LinearRegression())])

In [18]:
pipe["LR"].coef_

array([ 0.83817967,  0.12032096, -0.32135834,  0.36456599, -0.00170674,
       -0.04524631, -0.87994217, -0.82634126])

## **Que 8**

If we use `StandardScaler` for preprocessing and `LinearRegression` for fitting the model with training set used in previous question, what is the bias term predicted by the model?

**Note**: Use methods and objects with default parameters

a. 1.06855

b. 2.02067

c. 3.06855

d. -2.06855

**Correct Answer : b**

In [19]:
pipe[1].intercept_

2.020670313105614

# Graded Questions:

Total # of ques: 8

**Instructions**: For all graded questions, split the california housing dataset into train and test sets.

[Hint: use appropriate API with random_state=0, shuffle=False and test_size=0.2]

## **Que 9**

If we use `StandardScaler` for preprocessing and `LinearRegression` for fitting the model with training set used in previous question, what is the R squared value for predictions obtained using test set features?

Note: Use methods and objects with default parameters

a. 0.66051

b. 0.70623

c. 0.80623

d. 0.90623

**Correct Answer : a**

In [21]:
import sklearn.metrics as skm

In [22]:
y_test_pred=pipe.predict(X_test)

In [23]:
y_test_pred

array([1.31635001, 1.58760352, 1.11744227, ..., 0.14384532, 0.29510241,
       0.49760107])

In [24]:
skm.r2_score(y_test,y_test_pred)

0.6605140591531993

## **Que 10**

If we use `StandardScaler` for preprocessing and `LinearRegression` for fitting the model, what is the root mean squared error value for predictions obtained using test set features?

Note: Use methods and objects with default parameters

a. 0.8241

b. 0.9241

c. 0.7033

d. 0.6241

**Correct Answer : c**

In [28]:
skm.mean_squared_error(y_test,y_test_pred,squared=False)

0.7033383507521878

**Solution**


**Instructions** for next ques from Que 11 to 16

Split the data into training and test sets with random_state=0, shuffle=False and test_size=0.2 parameters.

Let, y_test= target label in test set of california housing dataset 

and y_pred=target labels obtained by the model using X_test

Then compute values of the following evaluation metrics


## **Que 11**
Explained Variance score

a) 0.66

b) 0.81

c) 0.73

d) 0.90

**Correct Answer: a**


In [29]:
skm.explained_variance_score(y_test,y_test_pred)

0.6605500501742703

## Que 12
Max Error

a. 0.18

b. 2.16

c. 0.66

d. 7.26

**Correct Answer: d**


In [30]:
skm.max_error(y_test,y_test_pred)

7.2604532929583705

## Que 13

Mean Absolute Error

a) 0.213

b) 0.516

c) 0.398

d) 0.422

**Correct Answer: b**


In [31]:
skm.mean_absolute_error(y_test,y_test_pred)

0.5168526993787042


## Que 14

Mean Squared Error

a) 0.038

b) 0.494

c) 0.506

d) 0.872

**Correct Answer: b**


In [32]:
skm.mean_squared_error(y_test,y_test_pred)

0.4946848356388075

**Instructions**: for Ques 15 to 16, perform SGD Regression on the given (scaled using StandardScalar()) dataset, using default hyperparameters and random_state=0,test_size=0.2, Shuffle=False

## **Que 15**

What is the bias term?

a) 2.0112

b) 1.0704

c) 0.0704

d) 3.0704

**Correct Answer: a**



## Que 16

What are the coefficients in predicted model?

a) [ 0.9268,  2.1200, -0.2552 ,  0.3310 , -0.0076,0.02969, -0.9130 , -0.8845]

b) [ 1.8268,  0.1200, -2.2552 ,  0.3310 , -0.0076,0.02969, -0.9130 , -2.8845]

c) [ 0.8404, 0.1123  , -0.4121,  0.2159, -0.0178,-0.01480, -0.8739, -0.8391]

d) [ 0.9268,  0.3200, -0.2552 ,  0.4310 , -0.0076,0.02969, -0.9130 , -1.8845]

**Correct Answer: c**

In [45]:
Xtrain,Xtest,ytrain,ytest =train_test_split(X,y,random_state=0,test_size=0.2,shuffle=False)

In [46]:
from sklearn.linear_model import SGDRegressor

In [47]:
pip=Pipeline(steps=[('sc',StandardScaler()),("sgd",SGDRegressor(random_state=0))])

In [48]:
pip.fit(X,y)

Pipeline(steps=[('sc', StandardScaler()),
                ('sgd', SGDRegressor(random_state=0))])

In [52]:
def sgd(Xtrain,Xtest,ytrain,ytest,sc):
    pip=Pipeline(steps=[('sc',sc),("sgd",SGDRegressor(random_state=0))])
    pip.fit(X,y)
    print(pip["sgd"].coef_)
    print(pip["sgd"].intercept_)
Xtrain,Xtest,ytrain,ytest =train_test_split(X,y,random_state=42,test_size=0.2,shuffle=False)
sgd(Xtrain,Xtest,ytrain,ytest,StandardScaler())

[  6.04030923   0.89490505 -10.70274305  10.64034252  -1.85306938
  65.35446596   5.6130849    3.91920251]
[2.85491534]
